e2etests/tests/routes.go

```diff
@@ -0,0 +1,276 @@
+		Eventually(func() error {
+			return openperouter.DaemonsetRolled(cs, routerPods)
+		}, 2*time.Minute, time.Second).ShouldNot(HaveOccurred())
+
+		removeLeafPrefixes(infra.LeafAConfig)
+		removeLeafPrefixes(infra.LeafBConfig)
+	})
+
+	BeforeEach(func() {
+		err := Updater.CleanButUnderlay()
+		Expect(err).NotTo(HaveOccurred())
+	})
+
+	AfterEach(func() {
+		err := Updater.CleanButUnderlay()
+		Expect(err).NotTo(HaveOccurred())
+	})
+
+	Context("with vnis", func() {
+		BeforeEach(func() {
+			err := Updater.Update(config.Resources{
+				VNIs: []v1alpha1.VNI{
+					vniRed,
+					vniBlue,
+				},
+			})
+			Expect(err).NotTo(HaveOccurred())
+
+			removeLeafPrefixes(infra.LeafAConfig)
+			removeLeafPrefixes(infra.LeafBConfig)
+		})
+
+		It("receives type 5 routes from the fabric", func() {
+			Contains := true
+			checkRouteForLeaf := func(leaf infra.Leaf, vni v1alpha1.VNI, mustContain bool, prefixes []string) {
+				Eventually(func() error {
+					for _, p := range routerPods {
+						exec := executor.ForPod(p.Namespace, p.Name, "frr")
+						evpn, err := frr.EVPNInfo(exec)
+						Expect(err).NotTo(HaveOccurred())
+						for _, prefix := range prefixes {
+							if mustContain && !evpn.ContainsType5RouteForVNI(prefix, leaf.VTEPIP, int(vni.Spec.VNI)) {
+								return fmt.Errorf("type5 route for %s - %s not found in %v in router %s", prefix, leaf.VTEPIP, evpn, p.Name)
+							}
+							if !mustContain && evpn.ContainsType5RouteForVNI(prefix, leaf.VTEPIP, int(vni.Spec.VNI)) {
+								return fmt.Errorf("type5 route for %s - %s found in %v in router %s", prefix, leaf.VTEPIP, evpn, p.Name)
+							}
+						}
+					}
+					return nil
+				}, 2*time.Minute, time.Second).ShouldNot(HaveOccurred())
+			}
+
+			By("announcing type 5 routes on VNI 100 from leafA")
+			changeLeafPrefixes(infra.LeafAConfig, leafAVRFRedPrefixes, emptyPrefixes)
+			checkRouteForLeaf(infra.LeafAConfig, vniRed, Contains, leafAVRFRedPrefixes)
+
+			By("announcing type5 route on VNI 200 from leafA")
+			changeLeafPrefixes(infra.LeafAConfig, leafAVRFRedPrefixes, leafAVRFBluePrefixes)
+			checkRouteForLeaf(infra.LeafAConfig, vniRed, Contains, leafAVRFRedPrefixes)
+			checkRouteForLeaf(infra.LeafAConfig, vniBlue, Contains, leafAVRFBluePrefixes)
+
+			By("announcing type5 route on both VNIs from leafB")
+			changeLeafPrefixes(infra.LeafBConfig, leafBVRFRedPrefixes, leafBVRFBluePrefixes)
+			checkRouteForLeaf(infra.LeafBConfig, vniRed, Contains, leafBVRFRedPrefixes)
+			checkRouteForLeaf(infra.LeafBConfig, vniBlue, Contains, leafBVRFBluePrefixes)
+
+			By("removing a route from leafA on vni 100")
+			changeLeafPrefixes(infra.LeafAConfig, emptyPrefixes, leafAVRFBluePrefixes)
+			checkRouteForLeaf(infra.LeafAConfig, vniRed, !Contains, leafAVRFRedPrefixes)
+			checkRouteForLeaf(infra.LeafAConfig, vniBlue, Contains, leafAVRFBluePrefixes)
+
+			By("removing a route from leafA on vni 200")
+			changeLeafPrefixes(infra.LeafAConfig, emptyPrefixes, emptyPrefixes)
```

Comment:
missing a check?

Reply:
I am going to add the 4 checks in every step, feels less error prone / easier to understand.